### PR TITLE
PoseEstimator interfacing with outer classes

### PR DESCRIPTION
**Pull request by Sh-Pe**
More PRs are coming soon

### CODE REVIEW

**Review by Yotam7 — APPROVED** [review]
Top Google result for gyro(used here in `gyroAngle`), we need to have a talk about shortening names
image

**Review by bossyshrimp68 — APPROVED** [review]

File: src/main/java/frc/robot/poseestimator/IOdometryEstimator.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import frc.robot.poseestimator.observations.OdometryObservation;
+
+public interface IOdometryEstimator {
+
+	void updateOdometry(OdometryObservation[] odometryObservation);
```
Comment (emilyyyylime) [inline src/main/java/frc/robot/poseestimator/IOdometryEstimator.java:10]:
```suggestion
	void updateOdometry(OdometryObservation[] odometryObservations);
```

File: src/main/java/frc/robot/poseestimator/IPoseEstimator.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import edu.wpi.first.math.geometry.Pose2d;
+
+
+public interface IPoseEstimator extends IVisionEstimator, IOdometryEstimator {
+
+	void resetPose(Pose2d newPose);
+
+	Pose2d getEstimatedPose();
+
+	Pose2d getEstimatedPoseAtTimeStamp(double timeStamp);
```
Comment (emilyyyylime) [inline src/main/java/frc/robot/poseestimator/IPoseEstimator.java:12]: Timestamp is usually one word 
```suggestion
	Pose2d getEstimatedPoseAtTimestamp(double timestamp);
```

File: src/main/java/frc/robot/poseestimator/observations/OdometryObservation.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package frc.robot.poseestimator.observations;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+
+public record OdometryObservation(SwerveModulePosition[] wheelsPositions, Rotation2d gyroAngle, double timestamp) {}
```
Comment (emilyyyylime) [inline src/main/java/frc/robot/poseestimator/observations/OdometryObservation.java:6]:
```suggestion
public record OdometryObservation(SwerveModulePosition[] wheelPositions, Rotation2d gyroAngle, double timestamp) {}
```

File: src/main/java/frc/robot/poseestimator/IPoseEstimator.java (hunk 2 above)
Comment (emilyyyylime) [reply src/main/java/frc/robot/poseestimator/IPoseEstimator.java:12]: In fact you used this same spelling in other files; so this is more consistent too

**Review by emilyyyylime — CHANGES_REQUESTED** [review]

**Review by emilyyyylime — APPROVED** [review]

**Review by drorta — DISMISSED** [review]

File: src/main/java/frc/robot/poseestimator/IOdometryEstimator.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import frc.robot.poseestimator.observations.OdometryObservation;
+
+public interface IOdometryEstimator {
+
+	void updateOdometry(OdometryObservation[] odometryObservations);
+
+	void resetOdometry(SwerveModulePosition[] wheelPositions, Rotation2d gyroAngle, Pose2d robotPose);
```
[the remark this answers is not in the document]
Comment (GilStein1) [reply src/main/java/frc/robot/poseestimator/IOdometryEstimator.java:12]: 2 different things, gyro might not be reset so its' rotation and the odometryPoses' rotation are not equal

File: src/main/java/frc/robot/poseestimator/IOdometryEstimator.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+	Pose2d getOdometryPose();
+
+	void resetHeadingOffset(Rotation2d newHeading);
```
[the remark this answers is not in the document]
Comment (GilStein1) [reply src/main/java/frc/robot/poseestimator/IOdometryEstimator.java:16]: new heading, the offset is calculated from the new robotHeading

File: src/main/java/frc/robot/poseestimator/IOdometryEstimator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (emilyyyylime) [reply src/main/java/frc/robot/poseestimator/IOdometryEstimator.java:12]: apostrphe's :3

File: src/main/java/frc/robot/poseestimator/IOdometryEstimator.java [hunk 6]
```diff
@@ -13,6 +13,6 @@ public interface IOdometryEstimator {
 
 	Pose2d getOdometryPose();
 
-	void resetHeadingOffset(Rotation2d newHeading);
+	void calculateHeadingOffsetFromRobotHeading(Rotation2d newHeading);
```
Comment (emilyyyylime) [inline src/main/java/frc/robot/poseestimator/IOdometryEstimator.java:16]: How about `setHeading`? I don't think it's that relevant for the callers of this function to know how it does what it does

**Review by drorta — APPROVED** [review]

**Review by GilStein1 — APPROVED** [review]
approve from Shahar (he opened the PR for me and now can't comment, he didn't write this code)

**Review by hillelv — APPROVED** [review]

**Review by emilyyyylime — APPROVED** [review]

**Review by aaa-png — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Pull request merged**